biomass/GEOCARBON/convert.py

```diff
@@ -0,0 +1,99 @@
+    outfile : str
+        the name of the file for the coarsened output
+    res : float, optional
+        the approximate resolution of the coarsen output in degrees
+    ntile : int, optional the number of coarse cells to process at a
+        time, routine will process a ntile x ntile subgrid. Increase
+        to run faster, decrease to reduce the peak memory.
+
+    """
+    with xr.open_dataset(filename) as ds:
+        ds = ds.rename({"latitude": "lat", "longitude": "lon"})
+        res0 = float(ds.lon.diff(dim="lon").mean())
+        n = int(round(res / res0))
+        if ntile is not None:
+            ds = ds.chunk({"lat": ntile * n, "lon": ntile * n})
+        c = ds.coarsen({"lat": n, "lon": n}, boundary="pad").mean()
+    return c
+
+
+if __name__ == "__main__":
+
+    # If local file does not exist, download
+    remote_source = "https://www.bgc-jena.mpg.de/geodb/projects/DataDnld.php"
+    local_source = "Forest_Aboveground_Biomassv3.nc"
+    download_stamp = time.strftime(
+        "%Y-%m-%d", time.localtime(os.path.getmtime(local_source))
+    )
+    generate_stamp = time.strftime("%Y-%m-%d")
+
+    # Coarsen
+    ds = CoarsenDataset(local_source, res=0.5)
+
+    # For some unknown reason, we need to re-create the biomass
+    # dataarray such that the encoding works out properly. I was
+    # having the problem that unit conversions didn't fail, but
+    # neither did they change the magnitudes of the data.
+    ds["biomass"] = xr.DataArray(
+        np.ma.masked_invalid(
+            ds["Forest_Aboveground_Biomass_v3"]
+            .to_numpy()
+            .reshape((1,) + ds["Forest_Aboveground_Biomass_v3"].shape)
+        ),
+        dims=("time", "lat", "lon"),
+    )
+    ds["biomass"].attrs = {"long_name": "above_ground_biomass", "units": "Mg ha-1"}
+    ds = ds.drop_vars("Forest_Aboveground_Biomass_v3")
+
+    # Having trouble with the time so let's just re-encode it
+    ds["time"] = [cftime.DatetimeNoLeap(2015, 6, 1)]
+    tb = np.asarray(
+        [
+            [cftime.DatetimeNoLeap(2000, 1, 1), cftime.DatetimeNoLeap(2010, 1, 1)],
+        ]
+    )
+    ds["time_bnds"] = xr.DataArray(tb, dims=("time", "nv"))
+    ds["time"].encoding["units"] = "days since 2000-01-01"
+    ds["time"].attrs["bounds"] = "time_bnds"
+
+    ds = ds.sortby(list(ds.sizes.keys()))
+    ds.attrs = {}
+    ds.attrs["title"] = (
+        "GEOCARBON: Towards an Operational Global Carbon Observing System"
+    )
+    ds.attrs["version"] = "3"
+    ds.attrs["institution"] = "GEOCARBON (European FP7 project)"
+    ds.attrs["source"] = (
+        "Files are available https://www.bgc-jena.mpg.de/geodb/projects/Home.php"
+    )
+    ds.attrs[
+        "history"
+    ] = f"""
+{download_stamp}: downloaded source from {remote_source};
+{generate_stamp}: coarsened to 0.5 degree resolution using https://github.com/rubisco-sfa/ILAMB-Data/blob/master/biomass/GEOCARBON/convert.py;"""
+
+    ds.attrs[
+        "references"
+    ] = """
+@ARTICLE{GEOCARBON,
```

Comment:
This wouldn't be an article would it? Also, is there an actual article you can reference? Here is my painstakingly vetted one from GFW:
```
@article{Cook-Patton2020,
author  = {Cook-Patton, Susan and Leavitt, Sara and Gibbs, David and Harris, Nancy and Lister, Kristine and Anderson-Teixeira, Kristina and Briggs, Russell and Chazdon, Robin and Crowther, Thomas and Ellis, Peter and Griscom, Heather and Herrmann, Valentine and Holl, Karen and Houghton, Richard and Larrosa, Cecilia and Lomax, Guy and Lucas, Richard and Madsen, Palle and Malhi, Yadvinder and Paquette, Alain and Parker, John and Paul, Keryn and Routh, Devin and Roxburgh, Stephen and Saatchi, Sassan and van den Hoogen, Johan and Walker, Wayne and Wheeler, Charlotte and Wood, Stephen and Xu, Liang and Griscom, Bronson},
title   = {Mapping carbon accumulation potential from global natural forest regrowth},
journal = {Nature},
year    = {2020},
volume  = {585},
pages   = {545--550},
doi     = {10.1038/s41586-020-2686-x}}"""
```

Reply:
I have not found a pub for this product :/ or instructions in how to reference it. I don't even know the institutions, I just put the project grant #. This is one of the most undocumented datasets we work with.

Reply:
Linking this here again since it has a few article references in it: https://datacore-gn.unepgrid.ch/geonetwork/srv/api/records/5e695176-266d-4697-bc06-c2d9196845b4